File: java/springmvc/src/main/java/com/lacunasoftware/pkisuite/controller/SignatureSessionPreDefRestCoreController.java
package com.lacunasoftware.pkisuite.controller;

import com.lacunasoftware.pkisuite.util.StorageMock;
import com.lacunasoftware.pkisuite.util.Util;
import com.lacunasoftware.restpki.*;
import org.springframework.stereotype.Controller;
import org.springframework.ui.Model;
import org.springframework.web.bind.annotation.*;

import java.util.UUID;
import java.util.List;
import java.util.ArrayList;

import javax.servlet.http.HttpServletResponse;

@Controller
public class SignatureSessionPreDefRestCoreController {

	private final String RETURN_URL = "http://localhost:60695/signature-session-predef-rest-core/complete";

	/**
	 * GET /signature-session-predef-rest-core
	 *
	 * This action initiates a signature session with REST PKI Core.
	 */
	@GetMapping("/signature-session-predef-rest-core")
	public String get(Model model, HttpServletResponse response) throws Exception {

		RestPkiService service = RestPkiServiceFactory.getService(Util.getRestPkiCoreOptions());

		CreateSignatureSessionRequest request = new CreateSignatureSessionRequest();
		request.setReturnUrl(RETURN_URL);
		request.disableDownloads(true);

		List<SignatureSessionDocumentToSign> docs = new ArrayList<>();

		for (int i = 0; i < 10; i++) {
			SignatureSessionDocumentToSign doc = new SignatureSessionDocumentToSign();
			doc.setFile(FileReference.FromFile(StorageMock.getBatchDocPath(i), String.format("%02d.pdf", i), "application/pdf"));
			doc.setSignatureType(SignatureTypes.PDF);
			docs.add(doc);
		}

		CreateSignatureSessionResponse sessionResponse = service.createSignatureSession(request, null, null, docs);

		// Render the signature-session-rest-core page (templates/signature-session-rest-core/index.html).
		model.addAttribute("redirectUrl", sessionResponse.getRedirectUrl());
		return "signature-session-predef-rest-core/index";
	}

	/**
	 * GET /signature-session-rest-core/webhook
	 *
	 * This action initiates a signature session with REST PKI Core using webhook.
	 */
	@GetMapping("/signature-session-predef-rest-core/webhook")
	public String usingWebhook(Model model, HttpServletResponse response) throws Exception {

		RestPkiService service = RestPkiServiceFactory.getService(Util.getRestPkiCoreOptions());

		CreateSignatureSessionRequest request = new CreateSignatureSessionRequest();
		request.setEnableBackgroundProcessing(true);
		request.disableDownloads(true);

		List<SignatureSessionDocumentToSign> docs = new ArrayList<>();

		for (int i = 0; i < 10; i++) {
			SignatureSessionDocumentToSign doc = new SignatureSessionDocumentToSign();
			doc.setFile(FileReference.FromFile(StorageMock.getBatchDocPath(i), String.format("%02d.pdf", i), "application/pdf"));
			doc.setSignatureType(SignatureTypes.PDF);
			docs.add(doc);
		}

		CreateSignatureSessionResponse sessionResponse = service.createSignatureSession(request, null, null, docs);

		// Render the signature-session-rest-core page (templates/signature-session-rest-core/index.html).
		model.addAttribute("redirectUrl", sessionResponse.getRedirectUrl());
		return "signature-session-predef-rest-core/index";
	}

	/**
	 * GET /ignature-session-rest-core/complete
	 *
	 * This action receives the form submission from the view. We'll call REST PKI to validate the
	 * authentication.
	 */
	@GetMapping("/signature-session-predef-rest-core/complete")
	public String complete(
		Model model,
		@RequestParam(value="signatureSessionId") String sessionId
	) throws Exception {
		// Get service
		RestPkiService service = RestPkiServiceFactory.getService(Util.getRestPkiCoreOptions());
		// Get session information
		SignatureSession session = service.getSignatureSession(UUID.fromString(sessionId));

		model.addAttribute("sessionStatus", session.getStatus());

		if (session.getStatus() == SignatureSessionStatus.COMPLETED){
			// If signature completed, get documents information
			List<SignatureSessionDocument> docs = session.getDocuments();
			model.addAttribute("documents", docs);
		}
		// Render the signature-session-rest-core page (templates/signature-session-rest-core/complete.html).
		return "signature-session-predef-rest-core/complete";
	}
}

File: dotnet/mvc/PkiSuiteAspNetMvcSample/Controllers/SignatureSessionPreDefRestCoreController.cs
using Lacuna.RestPki.Api;
using Lacuna.RestPki.Api.SignatureSessions;
using Lacuna.RestPki.Client;
using PkiSuiteAspNetMvcSample.Classes;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using System.Web;
using System.Web.Mvc;

namespace PkiSuiteAspNetMvcSample.Controllers
{
    public class SignatureSessionPreDefRestCoreController : Controller
    {
		#region Simple Flow

		// https://docs.lacunasoftware.com/articles/rest-pki/core/integration/signature-sessions/
		[HttpGet]
		public async Task<ActionResult> Index()
		{
			var restPkiService = Util.GetRestPkiService();

			var docs = new List<SignatureSessionDocumentToSign>();

			for (int i = 0; i < 10; i++)
			{
				SignatureSessionDocumentToSign doc = new SignatureSessionDocumentToSign();
				doc.File = FileReference.FromFile(StorageMock.GetBatchDocPath(i));
				doc.SignatureType = SignatureTypes.Pdf;
				docs.Add(doc);
			}

			var response = await restPkiService.CreateSignatureSessionAsync(new CreateSignatureSessionRequest()
			{
				ReturnUrl = "http://localhost:54123/SignatureSessionPreDefRestCore/Callback",
				DisableDownloads = true,
			}, documents: docs);

			return View(response);
		}

		// GET: SignatureSessionPreDefRestCore/Callback
		[HttpGet]
		public async Task<ActionResult> Callback(Guid signatureSessionId)
		{
			var restPkiService = Util.GetRestPkiService();
			var session = await restPkiService.GetSignatureSessionAsync(signatureSessionId);
			if (session.Status != SignatureSessionStatus.Completed)
			{
				return Redirect("/");
			}
			return View(session);
		}

		#endregion
	}
}